### PR TITLE
Concurrency, flip-policy, etc.

### PR DESCRIPTION
**Pull request by benediamond**
Implemented a number of great features inspired by @Akababa, including concurrency in the MCTS, flipping the policy vector for black's moves, and a number of other improvements in the MCTS (including the use of a proper tree, masking during first selection as opposed to during expansion, etc.).

### CODE REVIEW

**Comment by Akababa** [comment]
Nice! Quick update from my side: I actually found this [reddit comment](https://www.reddit.com/r/MachineLearning/comments/7ezd1q/research_project_leela_zero_a_community_open/dq8u57x/) which suggests good potential for C++, however a significant portion of the project would have to be rewritten. For now I'm just working on Python multiprocessing which should bypass the GIL and give better results for "low hanging fruit".

**Comment by benediamond** [comment]
Are you talking about pure C++ or Cython? In any case, I'm fairly familiar with how Leela Zero works. I even made some progress on a chess port of it before Samuel published his repository.

If you want to start a chess port of Leela Zero let me know, I can give you some early progress (I used stockfish's magic bitboards and move generation).

**Comment by Akababa** [comment]
Yeah that would be nice, Python is starting to give me a headache. Thanks!

**Comment by benediamond** [comment]
Ok, give me a day or so to dig it out and recover what I had done.

**Comment by Akababa** [comment]
Thanks. I'll also keep working on Python but find myself getting stuck since self-play is taking so long.

**Comment by Zeta36** [comment]
@benediamond, tell us if you get some convergence now in supervised learning!!

**Comment by benediamond** [comment]
I'm shocked to say that the results still don't look good. @Akababa, have you had any luck with supervised learning in your latest setup?

**Comment by Zeta36** [comment]
@benediamond, I'm sure the problem comes with the one-hot pieces configuration. Some time ago I could not make converge even your first try when you just changed my two planes integer ord() into one-hot planes.

**Comment by benediamond** [comment]
@Zeta36 but Google of course does the same thing.

**Comment by Zeta36** [comment]
Yes, I know but maybe we are doing it badly. Maybe we are creating the planes in a bad way. I don't know but as I tell you I could not make converge even your first try with one-hot pieces.

**Comment by Akababa** [comment]
@benediamond Unfortunately not :( It might simply be impossible to train our model with 9 million parameters (`model.summary()`) on a single GPU. We should probably try simplifications like removing all the history planes first.

**Comment by Zeta36** [comment]
I'd try first as a fast check to replace all the one-hot encoding planes by integer ord() integer planes. Just in case, @benediamond. As I told you I tried long time ago to train your first one-hot (one history) piece plane and it did not work neither.

**Comment by Akababa** [comment]
I will try that next, thanks. BTW do you think keras' embedding layer could be useful?

**Comment by benediamond** [comment]
I thought about the embedding layer, but I think it's unlikely to be useful. I think the standard use case is when you have a very large amount (i.e. millions) of categorical values, of which _some are related to each other_ and others are not. (for example, words or n-grams in a corpus). The embedding will map the vector space generated by such values into a much smaller-dimensional space, in such a way that similar things land near each other.

But in this case, we just have 6 (or rather 12) categories (namely the different types of pieces) and there aren't exactly strong relationships between them. and in any case 12 is sufficiently few features that there isn't a huge need for dimensionality reduction, I think.

**Comment by Zeta36** [comment]
Maybe @Akababa, but perhaps the first logic step is to know if the problem is or not in the one-hot embedding @benediamond used. If you replace all those planes by ord() ones and the model converges then we could try to find another way for implementing this embedding in the right way.

**Comment by Akababa** [comment]
I actually had some not-horrible results just now where I think I got my model to a point where it didn't give away pieces by training on FICS games with my evaluation function. Do you have any ideas on how to reduce the size of the numpy arrays? They're using too much i/o and memory..

**Comment by benediamond** [comment]
What did your losses look like during training, and what do you mean by evaluation function? and which numpy arrays are you talking about

**Comment by Akababa** [comment]
Take a look at my master https://github.com/Akababa/chess-alpha-zero/blob/master/src/chess_zero/env/chess_env.py#L98

Losses during training were around 2.5-3 although I halved the policy weight and it's most likely overfitted anyway. Basically the idea is that +-1 scores are too noisy so I replaced it with fraction of material. Also I think the Adam optimizer helped.

I think keras uses float32 numpy arrays while python lists are float64

**Comment by benediamond** [comment]
@Zeta36, After experimenting, I believe I was able to achieve losses comparable to yours by reverting features one by one. This leads me to believe that the difference between yours and mine may not be a bug, but rather a combination of the following factors:
1) I have been running with 19 residual layers (as in AlphaGo Zero), not 7.
2) I am using 4672 labels, not 1968, which will lead to higher policy log losses.
3) When you sample games at random instead of iterating through them during self-play, you are more likely to have repeated games, and thus to overfit to them.

I am not sure about these, but these could be food for thought as to why I am getting poorer performance.
